kvstore: Wait for kvstore watcher to exit

tgraf opened the pull request:
This fixes an overall watcher problem where kvstore watchers can continue running with events being emited and acted upon even after logic in code has decided to stop watching.

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/4945)
<!-- Reviewable:end -->

tgraf committed:
kvstore: Wait for kvstore watcher to exit

Also protect Stop() with a sync.Once to never close the channel twice.

Signed-off-by: Thomas Graf <thomas@cilium.io>

rlenglet commented on pkg/kvstore/events.go:
```diff
@@ -75,16 +81,27 @@ type Watcher struct {
 	stopped bool
+
+	// stopWait is the wait group to wait for watchers to exit gracefully
+	stopWait sync.WaitGroup
```
Both `stopWait` and `stopped` could be replaced with a single channel (e.g. named `stopped`) which would be closed by the watcher on exit. `Stop` could become:
```
w.mutex.Lock()
select {
    case <-w.stopWatch: // already stopped
        return
    default:
        close(w.stopWatch)
}
w.mutex.Unlock()
log.WithField(fieldWatcher, w).Debug("Stopped watcher")
<-w.stopped
```

tgraf replied:
Nice, I like it. Adjusted accordingly.

rlenglet reviewed (approved):
LGTM

aanm commented on pkg/kvstore/events.go:
```diff
@@ -107,11 +121,15 @@ func ListAndWatch(name, prefix string, chanSize int) *Watcher {
 // Stop stops a watcher previously created and started with Watch()
 func (w *Watcher) Stop() {
-	if w.stopped {
+	w.mutex.Lock()
```
I'm fine with this, but this is essentially a `sync.Do`. There's no need for the mutex in the Watcher struct.

tgraf replied:
Good point, will change in the next iteration.

tgraf replied:
fixed

tgraf committed:
consul: Add timeout to List() in Watch() to detect watcher stop

Due to the default timeout value in the blocking List(), a Consul watcher was
taking a long time to detect that the watcher must be stopped.

Signed-off-by: Thomas Graf <thomas@cilium.io>
